src/VisualStudio/Core/Def/Implementation/InheritanceMargin/InheritanceMarginViewMarginProvider.cs

```diff
@@ -0,0 +1,84 @@
+{
+    [Export(typeof(IWpfTextViewMarginProvider))]
+    [ContentType(ContentTypeNames.VisualBasicContentType)]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(nameof(InheritanceMarginViewMarginProvider))]
+    [MarginContainer(PredefinedMarginNames.Left)]
+    [Order(After = PredefinedMarginNames.Glyph)]
+    [TextViewRole(PredefinedTextViewRoles.Document)]
+    internal class InheritanceMarginViewMarginProvider : IWpfTextViewMarginProvider
+    {
+        private readonly IViewTagAggregatorFactoryService _tagAggregatorFactoryService;
+        private readonly IThreadingContext _threadingContext;
+        private readonly IStreamingFindUsagesPresenter _streamingFindUsagesPresenter;
+        private readonly IClassificationFormatMapService _classificationFormatMapService;
+        private readonly ClassificationTypeMap _classificationTypeMap;
+        private readonly IUIThreadOperationExecutor _operationExecutor;
+        private readonly IEditorFormatMapService _editorFormatMapService;
+        private readonly IAsynchronousOperationListenerProvider _listenerProvider;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public InheritanceMarginViewMarginProvider(
+            IThreadingContext threadingContext,
+            IStreamingFindUsagesPresenter streamingFindUsagesPresenter,
+            ClassificationTypeMap classificationTypeMap,
+            IClassificationFormatMapService classificationFormatMapService,
+            IUIThreadOperationExecutor operationExecutor,
+            IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IEditorFormatMapService editorFormatMapService,
+            IAsynchronousOperationListenerProvider listenerProvider)
+        {
+            _threadingContext = threadingContext;
+            _streamingFindUsagesPresenter = streamingFindUsagesPresenter;
+            _classificationTypeMap = classificationTypeMap;
+            _classificationFormatMapService = classificationFormatMapService;
+            _operationExecutor = operationExecutor;
+            _tagAggregatorFactoryService = tagAggregatorFactoryService;
+            _editorFormatMapService = editorFormatMapService;
+            _listenerProvider = listenerProvider;
+        }
+
+        public IWpfTextViewMargin? CreateMargin(IWpfTextViewHost wpfTextViewHost, IWpfTextViewMargin marginContainer)
+        {
+            var tagAggregator = _tagAggregatorFactoryService.CreateTagAggregator<InheritanceMarginTag>(wpfTextViewHost.TextView);
+            var editorFormatMap = _editorFormatMapService.GetEditorFormatMap(wpfTextViewHost.TextView);
+            var document = wpfTextViewHost.TextView.TextBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            var listner = _listenerProvider.GetListener(FeatureAttribute.InheritanceMargin);
+            if (document != null)
```

Comment:
i do not like that you're passing a Document snapshot (a point in time) to a UI element htat spans across document versions.  it looks like you might capture a Solution snapshot and keep all the data about it alive forever.

Reply:
It is just used to get the LanguageName & options service. So project/solution should be good. Will udpate this later